File: [0547][Medium][Friend_Circles]/Friend_Circles.py
# solution 1 greedy algorithm.
# leetcode time     cost : 220 ms
# leetcode memory   cost : 13.8 MB 
# Time  Complexity: O(n**2)
# Space Complexity: O(1)
# solution 3, UnionFind,
class Solution:
    def findCircleNum(self, M):
        def find(node):
            if circles[node] == node: return node
            root = find(circles[node])
            circles[node] = root
            return root
        
        n = len(M)
        circles = {x:x for x in range(n)}
        num = n
        for i in range(n):
            for j in range(i, n):
                if i != j and M[i][j] == 1 and find(i) != find(j):
                    circles[find(i)] = find(j)   
                    
        return sum([1 for k, v in circles.items() if k == v])
    
def main():
    M = [[1,1,0],[1,1,0],[0,0,1]]
    obj = Solution()
    res = obj.findCircleNum(M)
    print("return value is ",res)
    
if __name__ =='__main__':
    main()     

File: [0547][Medium][Friend_Circles]/Friend_Circles.java
// leetcode time     cost : 1 ms
// leetcode memory   cost : 42.5 MB 
// Time  Complexity: O(n**2)
// Space Complexity: O(n)
// solution 1, DFS,
class Solution {
    public void dfs(int[][] M, int[] visited, int i) {
        for (int j = 0; j < M.length; j++) {
            if (M[i][j] == 1 && visited[j] == 0) {
                visited[j] = 1;
                dfs(M, visited, j);
            }
        }
    }
    public int findCircleNum(int[][] M) {
        int[] visited = new int[M.length];
        int count = 0;
        for (int i = 0; i < M.length; i++) {
            if (visited[i] == 0) {
                dfs(M, visited, i);
                count++;
            }
        }
        return count;
    }
}

public class Friend_Circles {
    public static void main(String args[]) {
        int[][] M = {{1,1,0},{1,1,0},{0,0,1}}; // expect is 2,
        Solution Solution_obj = new Solution();
        int result = Solution_obj.findCircleNum(M);
        System.out.println("In java code,result is :" + result);
    }
}